### PR TITLE
fix(core): Update types for TypeScript nullability support

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: modules/benchmarks/src/util.ts [hunk 1]
```diff
@@ -60,7 +60,7 @@ export function profile(create: () => void, destroy: () => void, name: string) {
 function urlParamsToForm() {
   const regex = /(\w+)=(\w+)/g;
   const search = decodeURIComponent(location.search);
-  let match: any[];
+  let match: any[]|null;
```
Comment (snerks) [inline modules/benchmarks/src/util.ts]: Should `match` be initialised to `null` here?

Comment (mhevery) [inline modules/benchmarks/src/util.ts]: its not initialized, it is saying that it could be null.

Comment (DzmitryShylovich) [inline modules/benchmarks/src/util.ts]: I think he meant that the initial value in `undefined` so the type should be `let match: any[]|null|undefined;`

Comment (snerks) [inline modules/benchmarks/src/util.ts]: @DzmitryShylovich - thanks. That's what I failed to make clear.

File: modules/benchmarks/src/tree/util.ts [hunk 2]
```diff
@@ -13,10 +13,10 @@ export class TreeNode {
   children: TreeNode[];
 
   constructor(
-      public value: string, public depth: number, public maxDepth: number, public left: TreeNode,
-      public right: TreeNode) {
+      public value: string, public depth: number, public maxDepth: number,
+      public left: TreeNode|null, public right: TreeNode|null) {
     this.transitiveChildCount = Math.pow(2, (this.maxDepth - this.depth + 1)) - 1;
-    this.children = this.left ? [this.left, this.right] : [];
+    this.children = this.left ? [this.left, this.right !] : [];
```
Comment (tbosch) [inline modules/benchmarks/src/tree/util.ts]: change to `this.left && this.right`, then remove `this.right !`

File: packages/compiler-cli/integrationtest/test/source_map_spec.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ function getSourcePositionForStack(stack: string): {source: string, line: number
   const htmlLocations = stack
                             .split('\n')
                             // e.g. at View_MyComp_0 (...html:153:40)
-                            .map(line => /\((.*\.html):(\d+):(\d+)/.exec(line))
+                            .map(line => /\((.*\.html):(\d+):(\d+)/.exec(line) !)
```
Comment (tbosch) [inline packages/compiler-cli/integrationtest/test/source_map_spec.ts]: We should not need this as we filter afterwards to remove the `null` / `undefined`...

File: packages/compiler-cli/integrationtest/test/test_summaries.ts [hunk 4]
```diff
@@ -34,7 +34,8 @@ function main() {
       if (/.*\/node_modules\/.*/.test(fileName) && !/.*ngsummary\.json$/.test(fileName) &&
           !/package\.json$/.test(fileName)) {
         // Only allow to read summaries and package.json files from node_modules
-        return null;
+        // TODO (mhevery): Fix this. TypeScript.d.ts does not allow returning null.
```
Comment (tbosch) [inline packages/compiler-cli/integrationtest/test/test_summaries.ts]: Remove "TODO", add "Note: TypeScript API is not strictNullChecks compliant".

File: packages/compiler/src/view_compiler/view_compiler.ts [hunk 5]
```diff
@@ -644,8 +644,8 @@ class ViewBuilder implements TemplateAstVisitor, LocalResolver {
       return EventHandlerVars.event;
     }
     let currViewExpr: o.Expression = VIEW_VAR;
-    for (let currBuilder: ViewBuilder = this; currBuilder;
-         currBuilder = currBuilder.parent, currViewExpr = currViewExpr.prop('parent')) {
+    for (let currBuilder: ViewBuilder = this; currBuilder; currBuilder = currBuilder.parent,
```
Comment (tbosch) [inline packages/compiler/src/view_compiler/view_compiler.ts]: Add a "TODO" to fix this later...

File: packages/compiler/src/view_compiler/view_compiler.ts [hunk 6]
```diff
@@ -717,7 +717,7 @@ class ViewBuilder implements TemplateAstVisitor, LocalResolver {
       let compBuilder: ViewBuilder = this;
       while (compBuilder.parent) {
         compBuilder = compBuilder.parent;
-        compViewExpr = compViewExpr.prop('parent');
+        compViewExpr = compViewExpr.prop('parent').cast(o.DYNAMIC_TYPE);
```
Comment (tbosch) [inline packages/compiler/src/view_compiler/view_compiler.ts]: Add "TODO" to fix this later

File: packages/core/src/application_ref.ts [hunk 7]
```diff
@@ -299,7 +299,7 @@ export class PlatformRef_ extends PlatformRef {
         throw new Error('No ErrorHandler. Is platform module (BrowserModule) included?');
       }
       moduleRef.onDestroy(() => remove(this._modules, moduleRef));
-      ngZone.onError.subscribe({next: (error: any) => { exceptionHandler.handleError(error); }});
+      ngZone !.onError.subscribe({next: (error: any) => { exceptionHandler.handleError(error); }});
```
Comment (tbosch) [inline packages/core/src/application_ref.ts]: Add "TODO: use a new variable that is not null", or do the change.

File: packages/core/src/application_ref.ts [hunk 8]
```diff
@@ -500,7 +500,8 @@ export class ApplicationRef_ extends ApplicationRef {
     if (componentOrFactory instanceof ComponentFactory) {
       componentFactory = componentOrFactory;
     } else {
-      componentFactory = this._componentFactoryResolver.resolveComponentFactory(componentOrFactory);
+      componentFactory =
+          this._componentFactoryResolver.resolveComponentFactory(componentOrFactory) !;
```
Comment (tbosch) [inline packages/core/src/application_ref.ts]: `ComponentFactoryResolver.resolveComponentFactory` never returns null.

File: packages/core/src/change_detection/differs/iterable_differs.ts [hunk 9]
```diff
@@ -30,7 +30,7 @@ export interface IterableDiffer<V> {
    * @returns an object describing the difference. The return value is only valid until the next
    * `diff()` invocation.
    */
-  diff(object: NgIterable<V>): IterableChanges<V>;
+  diff(object: NgIterable<V>): IterableChanges<V>|null;
```
Comment (tbosch) [inline packages/core/src/change_detection/differs/iterable_differs.ts]: `object` should also be `NgIterable<V> | null`

File: packages/core/src/debug/debug_node.ts [hunk 10]
```diff
@@ -203,4 +199,4 @@ export function removeDebugNodeFromIndex(node: DebugNode) {
  *
  * @experimental All debugging apis are currently experimental.
  */
-export interface Predicate<T> { (value: T, index?: number, array?: T[]): boolean; }
```
Comment (tbosch) [inline packages/core/src/debug/debug_node.ts]: Keep `index` and `array` as optional and mark them as `@deprecated` as we never use them in any of our implementations, nor fill them.

File: packages/core/src/debug/debug_node.ts [hunk 11]
```diff
@@ -153,7 +149,7 @@ function _queryElementChildren(
     element: DebugElement, predicate: Predicate<DebugElement>, matches: DebugElement[]) {
   element.childNodes.forEach(node => {
     if (node instanceof DebugElement) {
-      if (predicate(node)) {
```
Comment (tbosch) [inline packages/core/src/debug/debug_node.ts]: Revert, see below

File: packages/core/src/debug/debug_node.ts [hunk 12]
```diff
@@ -165,7 +161,7 @@ function _queryNodeChildren(
     parentNode: DebugNode, predicate: Predicate<DebugNode>, matches: DebugNode[]) {
   if (parentNode instanceof DebugElement) {
     parentNode.childNodes.forEach(node => {
-      if (predicate(node)) {
```
Comment (tbosch) [inline packages/core/src/debug/debug_node.ts]: Revert, see below

File: packages/core/src/profile/wtf_impl.ts [hunk 13]
```diff
@@ -53,7 +53,9 @@ export function createScope(signature: string, flags: any = null): any {
   return events.createScope(signature, flags);
 }
 
-export function leave<T>(scope: Scope, returnValue?: T): T {
+export function leave<T>(scope: Scope): void;
```
Comment (tbosch) [inline packages/core/src/profile/wtf_impl.ts]: Why do we need this line? If we need it, then the next line should not use `?`

File: packages/core/src/util/decorators.ts [hunk 14]
```diff
@@ -341,7 +341,7 @@ export function makeParamDecorator(
       }
 
       parameters[index] = parameters[index] || [];
```
Comment (tbosch) [inline packages/core/src/util/decorators.ts]: either create a new variable that is non null for the value, or add a "TODO: create a new variable that is not null"

File: packages/core/src/view/element.ts [hunk 15]
```diff
@@ -61,13 +59,13 @@ export function elementDef(
...
     handleEvent?: ElementHandleEventFn, componentView?: ViewDefinitionFactory,
-    componentRendererType?: RendererType2): NodeDef {
+    componentRendererType?: RendererType2 | null): NodeDef {
   if (!handleEvent) {
     handleEvent = NOOP;
   }
   const {matchedQueries, references, matchedQueryIds} = splitMatchedQueriesDsl(matchedQueriesDsl);
-  let ns: string;
-  let name: string;
+  let ns: string = null !;
```
Comment (tbosch) [inline packages/core/src/view/element.ts]: I don't thinkg we need this...

File: packages/core/src/view/query.ts [hunk 16]
```diff
@@ -94,18 +94,18 @@ export function checkAndUpdateQuery(view: ViewData, nodeDef: NodeDef) {
     return;
   }
   let directiveInstance: any;
-  let newValues: any[];
+  let newValues: any[] = undefined !;
```
Comment (vicb) [inline packages/core/src/view/query.ts]: not needed ?

File: packages/core/src/view/refs.ts [hunk 17]
```diff
@@ -135,7 +135,7 @@ class ViewContainerRef_ implements ViewContainerData {
     let elDef = this._elDef.parent;
     while (!elDef && view) {
       elDef = viewParentEl(view);
-      view = view.parent;
+      view = view.parent !;
```
Comment (vicb) [inline packages/core/src/view/refs.ts]: either create a new variable that is non null for the value, or add a "TODO: create a new variable that is not null"

File: packages/core/src/view/services.ts [hunk 18]
```diff
@@ -281,44 +281,44 @@ function normalizeDebugBindingValue(value: any): string {
...
     let elView = view;
     while (elDef && (elDef.flags & NodeFlags.TypeElement) === 0) {
-      elDef = elDef.parent;
+      elDef = elDef.parent !;
     }
     if (!elDef) {
       while (!elDef && elView) {
-        elDef = viewParentEl(elView);
-        elView = elView.parent;
+        elDef = viewParentEl(elView) !;
```
Comment (vicb) [inline packages/core/src/view/services.ts]: either create a new variable that is non null for the value, or add a "TODO: create a new variable that is not null"

File: packages/core/src/view/services.ts [hunk 19]
```diff
@@ -408,14 +408,14 @@ function getRenderNodeIndex(viewDef: ViewDefinition, nodeIndex: number): number
   return renderNodeIndex;
 }
 
-function findHostElement(view: ViewData): ElementData {
+function findHostElement(view: ViewData): ElementData|null {
   while (view && !isComponentView(view)) {
-    view = view.parent;
+    view = view.parent !;
```
Comment (vicb) [inline packages/core/src/view/services.ts]: either create a new variable that is non null for the value, or add a "TODO: create a new variable that is not null"

File: packages/core/src/view/util.ts [hunk 20]
```diff
@@ -213,7 +215,7 @@ export function getParentRenderElement(view: ViewData, renderHost: any, def: Nod
 const VIEW_DEFINITION_CACHE = new WeakMap<any, ViewDefinition>();
 
 export function resolveViewDefinition(factory: ViewDefinitionFactory): ViewDefinition {
-  let value: ViewDefinition = VIEW_DEFINITION_CACHE.get(factory);
+  let value: ViewDefinition = VIEW_DEFINITION_CACHE.get(factory) !;
```
Comment (vicb) [inline packages/core/src/view/util.ts]: either create a new variable that is non null for the value, or add a "TODO: create a new variable that is not null"

File: packages/core/src/view/view_attach.ts [hunk 21]
```diff
@@ -10,13 +10,14 @@ import {ElementData, Services, ViewData} from './types';
...
 
 export function attachEmbeddedView(
-    parentView: ViewData, elementData: ElementData, viewIndex: number, view: ViewData) {
-  let embeddedViews = elementData.viewContainer._embeddedViews;
-  if (viewIndex == null) {
+    parentView: ViewData, elementData: ElementData, viewIndex: number | undefined | null,
+    view: ViewData) {
+  let embeddedViews = elementData.viewContainer !._embeddedViews;
+  if (viewIndex === null || viewIndex === undefined) {
     viewIndex = embeddedViews.length;
```
Comment (vicb) [inline packages/core/src/view/view_attach.ts]: either create a new variable that is non null for the value, or add a "TODO: create a new variable that is not null"

File: packages/core/test/application_ref_spec.ts [hunk 22]
```diff
@@ -92,7 +92,8 @@ export function main() {
          createRootEl();
          const modFactory = compiler.compileModuleSync(SomeModule);
          const module = modFactory.create(TestBed);
-         const cmpFactory = module.componentFactoryResolver.resolveComponentFactory(SomeComponent);
+         const cmpFactory =
```
Comment (vicb) [inline packages/core/test/application_ref_spec.ts]: should not be null

File: packages/core/test/change_detection/differs/default_iterable_differ_spec.ts [hunk 23]
```diff
@@ -309,7 +309,7 @@ export function main() {
 
         function modifyArrayUsingOperation(
             arr: number[], endData: any[], prev: number, next: number) {
-          let value: number = null;
+          let value: number = null !;
```
Comment (vicb) [inline packages/core/test/change_detection/differs/default_iterable_differ_spec.ts]: needed?

File: packages/core/test/linker/projection_integration_spec.ts [hunk 24]
```diff
@@ -234,7 +234,7 @@ export function main() {
     });
 
     it('should support moving non projected light dom around', () => {
-      let sourceDirective: ManualViewportDirective;
+      let sourceDirective: ManualViewportDirective = undefined !;
```
Comment (vicb) [inline packages/core/test/linker/projection_integration_spec.ts]: needed ?

File: packages/core/test/linker/security_integration_spec.ts [hunk 25]
```diff
@@ -75,7 +75,7 @@ function declareTests({useJit}: {useJit: boolean}) {
         });
 
         // should not throw for inputs starting with "on"
-        let cmp: ComponentFixture<SecuredComponent>;
+        let cmp: ComponentFixture<SecuredComponent> = undefined !;
```
Comment (vicb) [inline packages/core/test/linker/security_integration_spec.ts]: needed ?

File: packages/core/test/linker/source_map_integration_node_only_spec.ts [hunk 26]
```diff
@@ -26,15 +26,15 @@ export function main() {
     });
 
     function getErrorLoggerStack(e: Error): string {
-      let logStack: string;
-      getErrorLogger(e)(<any>{error: () => logStack = new Error().stack}, e.message);
+      let logStack: string = undefined !;
```
Comment (vicb) [inline packages/core/test/linker/source_map_integration_node_only_spec.ts]: needed ?

File: packages/core/test/linker/source_map_integration_node_only_spec.ts [hunk 27]
```diff
@@ -46,9 +46,9 @@ export function main() {
               .map(line => /\((.*\.ngfactory\.js):(\d+):(\d+)/.exec(line))
               .filter(match => !!match)
```
Comment (vicb) [inline packages/core/test/linker/source_map_integration_node_only_spec.ts]: Array.isArray(match) and remove ! ?

File: packages/core/test/linker/view_injector_integration_spec.ts [hunk 28]
```diff
@@ -184,19 +184,19 @@ class TestComp {
 
 export function main() {
   function createComponentFixture<T>(
-      template: string, providers: Provider[] = null, comp: Type<T> = null): ComponentFixture<T> {
+      template: string, providers?: Provider[] | null, comp?: Type<T>): ComponentFixture<T> {
     if (!comp) {
       comp = <any>TestComp;
```
Comment (vicb) [inline packages/core/test/linker/view_injector_integration_spec.ts]: either create a new variable that is non null for the value, or add a "TODO: create a new variable that is not null"

File: packages/core/testing/src/testing_internal.ts [hunk 29]
```diff
@@ -62,7 +62,7 @@ jsmBeforeEach(() => { testBed.resetTestingModule(); });
 
 function _describe(jsmFn: Function, ...args: any[]) {
   const parentRunner = runnerStack.length === 0 ? null : runnerStack[runnerStack.length - 1];
-  const runner = new BeforeEachRunner(parentRunner);
+  const runner = new BeforeEachRunner(parentRunner !);
```
Comment (vicb) [inline packages/core/testing/src/testing_internal.ts]: update BeforeEachRunner and remove !

File: packages/core/src/view/util.ts (hunk 20 above)
Comment (mhevery) [inline packages/core/src/view/util.ts]: fixed

File: packages/core/src/view/view_attach.ts (hunk 21 above)
Comment (mhevery) [inline packages/core/src/view/view_attach.ts]: fixed

File: packages/core/test/application_ref_spec.ts (hunk 22 above)
Comment (mhevery) [inline packages/core/test/application_ref_spec.ts]: fixed

File: packages/core/test/change_detection/differs/default_iterable_differ_spec.ts (hunk 23 above)
Comment (mhevery) [inline packages/core/test/change_detection/differs/default_iterable_differ_spec.ts]: fixed

File: packages/core/test/linker/projection_integration_spec.ts (hunk 24 above)
Comment (mhevery) [inline packages/core/test/linker/projection_integration_spec.ts]: yes, needed

File: packages/core/test/linker/security_integration_spec.ts (hunk 25 above)
Comment (mhevery) [inline packages/core/test/linker/security_integration_spec.ts]: yes

File: packages/core/test/linker/source_map_integration_node_only_spec.ts (hunk 26 above)
Comment (mhevery) [inline packages/core/test/linker/source_map_integration_node_only_spec.ts]: yes

File: packages/core/test/linker/source_map_integration_node_only_spec.ts (hunk 27 above)
Comment (mhevery) [inline packages/core/test/linker/source_map_integration_node_only_spec.ts]: fixed

File: packages/core/test/linker/view_injector_integration_spec.ts (hunk 28 above)
Comment (mhevery) [inline packages/core/test/linker/view_injector_integration_spec.ts]: fixed

File: packages/core/testing/src/testing_internal.ts (hunk 29 above)
Comment (mhevery) [inline packages/core/testing/src/testing_internal.ts]: fixed

**Review by vicb — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]

File: tools/public_api_guard/core/core.d.ts [hunk 30]
```diff
@@ -501,10 +501,10 @@ export declare function isDevMode(): boolean;
 
 /** @stable */
 export interface IterableChangeRecord<V> {
-    currentIndex: number;
-    item: V;
-    previousIndex: number;
-    trackById: any;
+    readonly currentIndex: number | null;
```
Comment (mhevery) [inline tools/public_api_guard/core/core.d.ts]: When it's removed

File: tools/public_api_guard/core/core.d.ts [hunk 31]
```diff
@@ -822,31 +822,31 @@ export declare abstract class Renderer2 {
     readonly abstract data: {
         [key: string]: any;
     };
-    destroyNode: (node: any) => void | null;
+    destroyNode: ((node: any) => void) | null;
     abstract addClass(el: any, name: string): void;
     abstract appendChild(parent: any, newChild: any): void;
     abstract createComment(value: string): any;
-    abstract createElement(name: string, namespace?: string): any;
+    abstract createElement(name: string, namespace?: string | null): any;
```
Comment (mhevery) [inline tools/public_api_guard/core/core.d.ts]: They don't but we should support falsy values.

**Pull request merged**